addons/stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -17,48 +17,117 @@ class StockPickingBatch(models.Model):
+        readonly=True, states={'draft': [('readonly', False)]})
 
+    @api.depends('company_id', 'picking_type_id')
+    def _compute_allowed_picking_ids(self):
+        for batch in self:
+            domain = [
+                ('company_id', '=', batch.company_id.id),
+                ('state', 'in', ['waiting', 'confirmed', 'assigned']),
+                '|',
+                ('batch_id', '=', False),
+                ('batch_id', '=', batch.id),
+            ]
+            if batch.picking_type_id:
+                domain += [('picking_type_id', '=', batch.picking_type_id.id)]
+            allowed_pickings = self.env['stock.picking'].search(domain)  # FIXME: could be replace by a read on id
+            batch.allowed_picking_ids = [(6, False, allowed_pickings.ids)]
+
+    @api.depends('picking_ids', 'picking_ids.move_lines')
+    def _compute_move_ids(self):
+        for batch in self:
+            batch.move_ids = batch.picking_ids.move_lines
+            batch.show_check_availability = any(m.state != 'assigned' for m in batch.move_ids)
+
+
+    # -------------------------------------------------------------------------
+    # CRUD
+    # -------------------------------------------------------------------------
     @api.model
     def create(self, vals):
         if vals.get('name', '/') == '/':
             vals['name'] = self.env['ir.sequence'].next_by_code('picking.batch') or '/'
-        return super(StockPickingBatch, self).create(vals)
+        res = super().create(vals)
+        res._sanity_check()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('picking_type_id'):
+            self._sanity_check()
+        elif vals.get('picking_ids'):
+            batch_without_picking_type = self.filtered(lambda batch: not batch.picking_type_id)
+            if batch_without_picking_type:
+                picking_command = vals.get('picking_ids')[0]
+                picking_id = picking_command[2] and picking_command[2][0]
+                picking = self.env['stock.picking'].browse(picking_id)
+                batch_without_picking_type.picking_type_id = picking.picking_type_id.id
+                self._sanity_check()
+        return res
+
+    def unlink(self):
+        if any(batch.state != 'draft' for batch in self):
+            raise UserError(_("You can only delete draft batch transfers."))
+        return super().unlink()
 
-    def confirm_picking(self):
+    # -------------------------------------------------------------------------
+    # Action methods
+    # -------------------------------------------------------------------------
+    def action_confirm(self):
+        """Sanity checks, assign the pickings and mark the batch as confirmed."""
+        self.ensure_one()
+        if not self.picking_ids:
+            raise UserError(_("You have to set some pickings to batch."))
+        self.picking_ids.action_assign()
```

Comment:
self.picking_ids.action_confirm()

Reply:
@Whenrow  As `action_confirm` is called in `action_assign`, it's better to keep `action_assign` here I think.

Reply:
Still would use action_confirm() as Willy said. Since it would keep the same behavior than picking them self https://github.com/odoo/odoo/blob/13.0/addons/stock/models/stock_picking.py#L615

Reply:
also action_assign() on batch did not exist before that's why action_assign was call previously